File: 347-top-k-frequent-elements/347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        
        map <int,int> mp;
        vector<int> ams;
        
        for(auto a:nums)
        {
            mp[a]=mp[a]+1;
            
        }
        // for(auto i:mp){
        //     cout<<i.first<<" "<<i.second<<endl;
        // }
        
        priority_queue <pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> pq;
        for(auto i:mp)
        {
            pq.push(make_pair(i.second,i.first));
            
            if(pq.size()>k) pq.pop();
        }
        
        while(pq.size()>0)
        {
            
            ams.push_back(pq.top().second);
            pq.pop();
        } 
    
    return ams;
    }
};

File: 347-top-k-frequent-elements/347-top-k-frequent-elements.java
class Solution {
    class pair implements Comparable<pair>{
        int num,fre;
        pair(int a,int b){
            num=a;
            fre=b;
        }
        
        public int compareTo(pair b){
            return this.fre-b.fre;
        }
    }
    public int[] topKFrequent(int[] nums, int k)
    {
        int n=nums.length;
        Queue<pair> pq=new PriorityQueue<>();
        Arrays.sort(nums);
        
        for(int i=0;i<n;i++){
            int count=1;
            while(i<n-1 && nums[i]==nums[i+1]){
                count++;
                i++;
            }
            pq.add(new pair(nums[i],count));
            
            if(pq.size()>k) pq.remove();
        }
        
        int ans[]=new int[k];
        int i=k-1;
        while(!pq.isEmpty()){
            ans[i--]=pq.remove().num;
        }
        return ans;
    }
}